suite/tests/client-interface/cleancallregs.dll.c

```diff
@@ -0,0 +1,270 @@
+/* **********************************************************
+ * Copyright (c) 2017 ARM Limited. All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of ARM Limited nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL ARM LIMITED OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Marshalling the arguments to a clean call is non-trivial when the arguments
+ * are passed in registers and the values in some registers need to be permuted
+ * (and we wish to do this efficiently). This test checks some interesting cases.
+ *
+ * It makes no difference what app is used. The first basic block encountered will
+ * be instrumented with a sequence of clean calls followed by a call via finish()
+ * to dr_exit_process(), so no app code is executed.
+ *
+ * This test can easily be ported to a new architecture. There is also the
+ * X86-specific test "cleancall".
```

Comment:
I'm worried that you're considering the existing test to be completely "x86-specific": it tests handling faults in clean call args, which is cross-platform; different states of condition codes, which is cross-platform; clean call args that use the stack pointer, which are complex to handle -- even on aarch once args don't fit in registers.  If adding this new test is going to mean that it's less likely that we ever port the old test, maybe it's better to merge the tests here into the existing one and get it working on arm?